x-pack/plugins/ml/public/application/services/elastic_models_service.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { type ModelDefinitionResponse } from '@kbn/ml-trained-models-utils';
+import { type TrainedModelsApiService } from './ml_api_service/trained_models';
+
+export type ElserVersion = 1 | 2;
+
+export class ElasticModels {
+  constructor(private readonly trainedModels: TrainedModelsApiService) {}
+
+  /**
+   * Provides an ELSER model name and configuration for download based on the current cluster architecture.
+   * The current default version is 2. If running on Cloud it returns the Linux x86_64 optimized version.
+   * If any of the ML nodes run a different OS rather than Linux, and the CPU architecture isn't x86_64,
+   * a portable version of the model is returned.
+   */
+  public async getELSER(options?: { version?: ElserVersion }): Promise<ModelDefinitionResponse> {
+    const response = await this.trainedModels.getTrainedModelDownloads();
+
+    let requestedModel: ModelDefinitionResponse | undefined;
+    let recommendedModel: ModelDefinitionResponse | undefined;
+    let defaultModel: ModelDefinitionResponse | undefined;
+
+    for (const model of response) {
+      if (options?.version === model.version) {
+        requestedModel = model;
+        if (model.recommended) {
```

Comment:
It appears this inner if statement isn't needed. 
if this is the requested model, it'll be returned over the others, so it doesn't matter if it is also the recommended one.

Seeing as the requested model is the highest priority, we could break out of the loop and just return it once it has been found.
The loop could even be broken up deal with the requested model first:
```ts
  public async getELSER(options?: { version?: ElserVersion }): Promise<ModelDefinitionResponse> {
    const response = await this.trainedModels.getTrainedModelDownloads();

    let defaultModel: ModelDefinitionResponse | undefined;

    if (options?.version !== undefined) {
      // find the requested model, otherwise throw error
      const requestedModel = response.find((model) => model.version === options.version);
      if (requestedModel) {
        return requestedModel;
      }
      throw new Error('Requested model not found');
    }

    // find the recommended model, otherwise the default model
    for (const model of response) {
      if (model.recommended) {
        return model;
      }
      if (model.default) {
        defaultModel = model;
      }
    }

    if (defaultModel) {
      return defaultModel;
    } else {
      throw new Error('Requested model not found');
    }
  }
}
```

Reply:
> It appears this inner if statement isn't needed.
> if this is the requested model, it'll be returned over the others, so it doesn't matter if it is also the recommended one.

It's actually needed, because there might be several versions of the requested models, and only one of them is recommended. E.g. the user requests version `2`, but we have a portable version of this model and platform optimized, which might be recommended. Hence we need to override the requested model in this case. You can check the unit tests for this function for this particular use-case.

Reply:
ah ok, cool